### PR TITLE
React16

### PR DESCRIPTION
**Pull request by EdwinGuzman**
This fixes [SIMPLE-1964](https://jira.nypl.org/browse/SIMPLY-1964).

Creating a draft first because the tests for `Lanes-tests` need to be updated/fixed but that will be refactored by PR #233 . For now that suite is skipped.

I will soon write documentation on the update from enzyme 2 to 3 because that gave me the most trouble and headaches. There are minor API changes but it affects how we write tests - some are bugs found in Enzyme that haven't fully been fixed and some are just "better" ways to handle DOM lookups.

I'd like to revisit all the tests again but this may not be the PR for that.

### CODE REVIEW

File: src/components/ClassificationsForm.tsx [hunk 1]
```diff
@@ -167,7 +167,7 @@ export default class ClassificationsForm extends React.Component<Classifications
             </div>
           }
         />
-        <Button callback={this.submit} content="Save" />
+        <Button callback={this.submit} content="Save" className="classification-save-btn" />
```
Comment (adriana-alter) [inline src/components/ClassificationsForm.tsx]: Rationale for adding the className?

Comment (EdwinGuzman) [reply src/components/ClassificationsForm.tsx]: For a test, but it seems I fixed the test without the need for this so I'll remove.

**Comment by adriana-alter** [comment]
This looks good to me (except for the new merge conflicts from my Lanes branch--sorry about that!)

**Comment by EdwinGuzman** [comment]
No, that's fine. I was waiting for the conflicts so I can continue updating this branch.

File: src/components/__tests__/InputList-test.tsx [hunk 2]
```diff
@@ -160,20 +162,20 @@ describe("InputList", () => {
   });
 
   it("adds an autocompleted item", () => {
-    let languageSetting = {...setting, ...{format: "language-code"}};
+    let languageSetting = { ...setting, format: "language-code" };
```
Comment (EdwinGuzman) [inline src/components/__tests__/InputList-test.tsx]: This was an unnecessary spread syntax here.

File: src/components/__tests__/Lanes-test.tsx [hunk 3]
```diff
@@ -67,8 +67,9 @@ describe("Lanes", () => {
   };
 
   const getTopLevelLanes = () => {
-    let sidebar = wrapper.find(".lanes-sidebar");
-    let topLevelLanes = sidebar.children(DragDropContext).children(Draggable).children("div");
+    const sidebar = wrapper.find(".lanes-sidebar");
+    // let topLevelLanes = sidebar.children(DragDropContext).children(Draggable).children("div");
+    const topLevelLanes = sidebar.find("ul.droppable").at(0).children();
```
Comment (EdwinGuzman) [inline src/components/__tests__/Lanes-test.tsx]: This was the most annoying thing to update because now `.children()` doesn't actually return what you expect.

File: src/components/__tests__/LanesSidebar-test.tsx [hunk 4]
```diff
@@ -285,54 +287,123 @@ describe("LanesSidebar", () => {
...
+    //        top
+    //       /   \
+    //      1     4
+    //     / \
+    //    2   5
+    //   /   / \
+    //  3   6   7
+    //      |
+    //      8
+    /** lane  8 */
```
Comment (EdwinGuzman) [inline src/components/__tests__/LanesSidebar-test.tsx]: There was no test for lane 8.

File: src/components/__tests__/LanesSidebar-test.tsx [hunk 5]
```diff
@@ -23,10 +23,10 @@ describe("LanesSidebar", () => {
     custom_list_ids: [2], inherit_parent_restrictions: false
```
Comment (EdwinGuzman) [inline src/components/__tests__/LanesSidebar-test.tsx]: This file was the most complicated to update so more eyes on it is appreciated.

**Comment by EdwinGuzman** [comment]
@aslagle @adriana-alter ready for review now.

**Comment by EdwinGuzman** [comment]
Also, I wanted to use the new `ref` API but it's too big for some cases that it could be its own PR (along with other React 16 changes).

File: src/components/__tests__/LanesSidebar-test.tsx [hunk 6]
```diff
@@ -23,10 +23,10 @@ describe("LanesSidebar", () => {
...
   };
   let lanesData: LaneData[];
-
   const getTopLevelLanes = () => {
-    return wrapper.children(DragDropContext).children(Draggable).children("div");
```
Comment (adriana-alter) [inline src/components/__tests__/LanesSidebar-test.tsx]: Can you put in a comment saying what this returns now (since we're not used to the counterintuitive new `children()` implementation yet!)?

File: src/components/__tests__/LanesSidebar-test.tsx [hunk 7]
```diff
@@ -49,56 +49,57 @@ describe("LanesSidebar", () => {
   });
   it("renders create top-level lane link", () => {
     let create = wrapper.find(".lanes-sidebar > div").find(Link).at(0);
-    expect(create.hasClass("create-lane")).to.be.true;
+    expect(create.render().hasClass("create-lane")).to.be.true;
```
Comment (adriana-alter) [inline src/components/__tests__/LanesSidebar-test.tsx]: What's the advantage of calling `render()` on these?  It looks to me like it would return the same thing.  Just a matter of best practices...?

File: src/components/__tests__/LanesSidebar-test.tsx [hunk 8]
```diff
@@ -49,56 +49,57 @@ describe("LanesSidebar", () => {
...
     topLevelLanes = getTopLevelLanes();
-    topLane1 = topLevelLanes.at(0);
-    topLane2 = topLevelLanes.at(1);
-    subLane1 = topLane1.find(Lane).find("> div");
-    expect(topLane1.hasClass("active")).to.be.true;
-    expect(topLane2.hasClass("active")).to.be.false;
-    expect(subLane1.hasClass("active")).to.be.false;
+    topLane1 = topLevelLanes.at(0).find("div").at(0);
+    topLane2 = topLevelLanes.at(1).find("div").at(0);
+    subLane1 = topLane1.find("li > div");
```
Comment (adriana-alter) [inline src/components/__tests__/LanesSidebar-test.tsx]: I think the previous `.find(Lane)` implementation is a lot more readable, since it's not obvious that `li > div` is a sublane; if there are advantages to doing it this way instead (which I'm willing to believe!), how about a comment?

File: src/components/__tests__/LanesSidebar-test.tsx [hunk 9]
```diff
@@ -176,56 +177,57 @@ describe("LanesSidebar", () => {
...
     expect(subLane1.text()).to.contain("Sublane 1");
     expect(subLane1.text()).to.contain("(3)");
     expect(subLane1.length).to.equal(1);
     let subLane1Expand = expectCollapsed(subLane1);
 
     // top lane 2 has no sublanes
-    let topLane2Sublanes = topLane2.find(Lane).find("> div");
-    let topLane2Draggables = topLane2.children(Droppable).children(Draggable).children("div");
+    let topLane2Sublanes = topLane2.find("li > div");
+    let topLane2Draggables = topLane2.find(Droppable).find(Draggable);
```
Comment (adriana-alter) [inline src/components/__tests__/LanesSidebar-test.tsx]: I think you could use the `allChildren` and `draggableChildren` helper functions for some of these.

File: src/components/__tests__/LanesSidebar-test.tsx [hunk 10]
```diff
@@ -285,54 +287,123 @@ describe("LanesSidebar", () => {
...
+    lane1Children = allChildren(lane1);
+    // The fourth child is the 8th lane since all lanes are expanded.
+    let sublane8 = lane1Children.at(4);
+    expect(isDraggable(sublane8)).to.be.false;
+    expand(sublane8);
+    expect(sublane8.text()).to.contain("sublane 8");
+    // Sublane 8 doesn't have children.
+    expect(allChildren(sublane8.render()).length).to.equal(0);
+    expect(draggableChildren(sublane8.render()).length).to.equal(0);
   });
```
Comment (adriana-alter) [inline src/components/__tests__/LanesSidebar-test.tsx]: I like all the new trees!

File: src/components/__tests__/ClassificationsForm-test.tsx [hunk 11]
```diff
@@ -89,7 +89,7 @@ describe("ClassificationsForm", () => {
 
     it("should not allow you to submit if you didn't select an audience or a fiction classification",
       () => {
-        let button = wrapper.find("button").findWhere(button => button.text() === "Save");
+        let button = wrapper.find("button").findWhere(button => button.text() === "Save").at(0);
```
[the remark this answers is not in the document]
Comment (adriana-alter) [reply src/components/__tests__/ClassificationsForm-test.tsx]: If it's a `Button`, it should have a className you could find it by instead...

File: src/components/__tests__/ClassificationsForm-test.tsx [hunk 12]
```diff
@@ -337,13 +340,14 @@ describe("ClassificationsForm", () => {
       let button = wrapper.find(WithRemoveButton);
       let onRemove = button.props().onRemove;
       onRemove();
+      wrapper.update();
```
[the remark this answers is not in the document]
Comment (adriana-alter) [reply src/components/__tests__/ClassificationsForm-test.tsx]: It seems weird that the new version of Enzyme would require us to manually do things that the old one did automatically; maybe this is an indication that there's a different pattern we should be using?

File: src/components/__tests__/ClassificationsForm-test.tsx (hunk 11 above)
[the remark this answers is not in the document]
Comment (EdwinGuzman) [reply src/components/__tests__/ClassificationsForm-test.tsx]: You'd still have to do the same thing if it's a React class, even if you search by class name.

File: src/components/__tests__/ClassificationsForm-test.tsx (hunk 12 above)
[the remark this answers is not in the document]
Comment (EdwinGuzman) [reply src/components/__tests__/ClassificationsForm-test.tsx]: This is a pattern I saw often. There could be a different one as well.

File: src/components/__tests__/LanesSidebar-test.tsx (hunk 7 above)
Comment (EdwinGuzman) [reply src/components/__tests__/LanesSidebar-test.tsx]: I think I explained why in a different file, but the component isn't fully rendered and `hasClass` won't work by itself, _in most cases_.

File: src/components/__tests__/LanesSidebar-test.tsx (hunk 8 above)
Comment (EdwinGuzman) [reply src/components/__tests__/LanesSidebar-test.tsx]: I agree but `find(Lane)` doesn't always work since the component is already mounted. Finding by class name would be better for a shallow rendering of the component, so I tried to stick to classes and elements rather than React class names.

File: src/components/__tests__/ClassificationsForm-test.tsx (hunk 11 above)
[the remark this answers is not in the document]
Comment (aslagle) [reply src/components/__tests__/ClassificationsForm-test.tsx]: This is annoying. Would `hostNodes` be more reliable since we don't have to assume the order of the DOM element and React component?

File: src/components/__tests__/CustomLists-test.tsx [hunk 13]
```diff
@@ -205,6 +205,10 @@ describe("CustomLists", () => {
     expect(descendingButton.prop("checked")).to.equal(false);
 
     descendingButton.find("input").simulate("change");
+    radioButtons = wrapper.find(EditableInput);
```
Comment (aslagle) [inline src/components/__tests__/CustomLists-test.tsx]: so how come this works without `hostNodes` or `at`? Because it's already a list, and each thing in the list now has two things in it?

File: src/components/__tests__/ResetAdobeId-test.tsx [hunk 14]
```diff
@@ -125,27 +125,29 @@ describe("ResetAdobeId", () => {
...
     it("should enable the submit button when the checkbox is checked", () => {
       const input = wrapper.find(".reset-adobe-id input");
-      const button = wrapper.find(".reset-adobe-id").find(Button);
+      let button = wrapper.find(".reset-adobe-id").find(Button);
 
       expect(wrapper.state().checked).to.equal(false);
 
       input.simulate("change");
-      wrapper.update();
+      // wrapper.update();
```
Comment (aslagle) [inline src/components/__tests__/ResetAdobeId-test.tsx]: It is pretty weird that `wrapper.update` needed to be added some places but removed others....

Comment (EdwinGuzman) [reply src/components/__tests__/ResetAdobeId-test.tsx]: Forgot to remove this. No instance method was called directly and the state/prop were not updated directly. So if I understand it correctly, it's not needed for this case and re-finding the DOM is what is needed instead.

File: src/components/__tests__/ClassificationsForm-test.tsx (hunk 11 above)
[the remark this answers is not in the document]
Comment (EdwinGuzman) [reply src/components/__tests__/ClassificationsForm-test.tsx]: Correct, it'll be the first or only node retrieved (if there are multiple).

File: src/components/__tests__/CustomLists-test.tsx (hunk 13 above)
Comment (EdwinGuzman) [reply src/components/__tests__/CustomLists-test.tsx]: Because you're not trying to find by class name.

File: src/components/__tests__/LanesSidebar-test.tsx (hunk 9 above)
Comment (EdwinGuzman) [reply src/components/__tests__/LanesSidebar-test.tsx]: It works 👌

File: package.json [hunk 15]
```diff
@@ -32,14 +32,14 @@
     "library-simplified-reusable-components": "1.3.7",
     "numeral": "^2.0.6",
     "opds-feed-parser": "0.0.14",
-    "opds-web-client": "0.2.0",
+    "opds-web-client": "0.2.2",
```
Comment (aslagle) [inline package.json]: Would you mind updating this to 0.2.3 so I don't have to make a separate PR just for that? I published it yesterday.

File: src/components/__tests__/SitewideSettingEditForm-test.tsx [hunk 16]
```diff
@@ -173,23 +173,26 @@ describe("SitewideSettingEditForm", () => {
     let fillOutFormFields = () => {
       wrapper.setProps({ item: settingData });
       let input = wrapper.find("input[name='value']");
-      let inputElement = input.get(0);
+      // Fixed message - Cannot add property value, object is not extensible
```
Comment (aslagle) [inline src/components/__tests__/SitewideSettingEditForm-test.tsx]: I don't understand this comment - is this why you have to use `getDOMNode`?

**Comment by aslagle** [comment]
I still don't totally understand some of the enzyme changes but I think this looks good overall.

File: package.json (hunk 15 above)
Comment (EdwinGuzman) [reply package.json]: Will do.

File: src/components/__tests__/SitewideSettingEditForm-test.tsx (hunk 16 above)
Comment (EdwinGuzman) [reply src/components/__tests__/SitewideSettingEditForm-test.tsx]: Yes, I'll update the comment. Using `get(0)` for input elements has been updated to `getDOMNode()`.

File: src/components/__tests__/CustomListsSidebar-test.tsx [hunk 17]
```diff
@@ -34,7 +34,7 @@ describe("CustomListsSidebar", () => {
   });
 
   it("renders a sidebar with a header and a create button", () => {
-    expect(wrapper.hasClass("custom-lists-sidebar")).to.be.true;
+    expect(wrapper.render().hasClass("custom-lists-sidebar")).to.be.true;
```
[the remark this answers is not in the document]
Comment (adriana-alter) [reply src/components/__tests__/CustomListsSidebar-test.tsx]: That is...annoying.  Any reason not to do it as `expect(wrapper.find(".custom-lists-sidebar").length` instead?

**Comment by adriana-alter** [comment]
This looks good!  (Well, much of the new Enzyme stuff does not look good.  But there's nothing we can do about that.)

File: package.json (hunk 15 above)
Comment (EdwinGuzman) [reply package.json]: Funny story, using 0.2.3 outputs errors (+ a few more not in the screenshot):
Screen Shot 2019-06-04 at 1 58 23 PM

Comment (EdwinGuzman) [reply package.json]: I'll investigate and see if I can fix it in this PR or if not make another PR.

Comment (aslagle) [reply package.json]: Oh weird. It might have to do with adding a subtitle to BookData, but I thought I made that optional. If it's not easy to fix I'll do a separate PR, but didn't want to conflict with this one.

Comment (EdwinGuzman) [reply package.json]: Oh you know what it is, both the client and parser need to be updated because the `OPDSEntry` interface is retrieved directly from the parser module. It raised other errors but fixed.

Comment (aslagle) [reply package.json]: Oh right! I forgot that was separately imported here.

File: src/__tests__/editorAdapter-test.ts [hunk 18]
```diff
@@ -21,13 +22,13 @@ describe("editorAdapter", () => {
...
       identifiers: [],
       links: [
-        { href: "hide", rel: "http://librarysimplified.org/terms/rel/hide", type: "type", title: "title" },
-        { href: "restore", rel: "http://librarysimplified.org/terms/rel/restore", type: "type", title: "title" },
-        { href: "refresh", rel: "http://librarysimplified.org/terms/rel/refresh", type: "type", title: "title" },
-        { href: "edit", rel: "edit", type: "type", title: "title" },
-        { href: "issues", rel: "issues", type: "type", title: "title" },
-        { href: "change-cover", rel: "http://librarysimplified.org/terms/rel/change_cover", type: "type", title: "title" },
-        { href: "cover", rel: "http://opds-spec.org/image", type: "image/png", title: "title" },
+        { href: "hide", rel: "http://librarysimplified.org/terms/rel/hide", type: "type", title: "title", role: "role" },
```
Comment (EdwinGuzman) [inline src/__tests__/editorAdapter-test.ts]: @aslagle I thought `role` was optional but not having it here threw errors...

File: src/editorAdapter.ts [hunk 19]
```diff
@@ -102,7 +95,7 @@ export default function adapter(data: OPDSEntry): BookData {
     title: data.title,
     authors: authors,
     contributors: data.contributors,
-    subtitle: subtitle,
+    subtitle: data.subtitle,
```
Comment (EdwinGuzman) [inline src/editorAdapter.ts]: Removed the previous way to retrieve the subtitle since that is now part of the OPDSEntry object.

File: package.json [hunk 20]
```diff
@@ -31,8 +31,8 @@
     "isomorphic-fetch": "^2.2.1",
     "library-simplified-reusable-components": "1.3.7",
     "numeral": "^2.0.6",
-    "opds-feed-parser": "0.0.14",
-    "opds-web-client": "0.2.2",
+    "opds-feed-parser": "0.0.16",
```
Comment (EdwinGuzman) [inline package.json]: The 0.0.15 bump introduced the role attribute in the OPDSLink object.

File: src/__tests__/editorAdapter-test.ts (hunk 18 above)
Comment (aslagle) [reply src/__tests__/editorAdapter-test.ts]: We might want to fix that in the parser, I think it generally always includes everything but sets some things to undefined.

File: src/editorAdapter.ts (hunk 19 above)
Comment (aslagle) [reply src/editorAdapter.ts]: Wow I hadn't realized this was here :)

Thanks for updating it.
